src/features/upload-file/upload-file.feature.ts

```diff
@@ -1,128 +1,203 @@
+        const oldBucket = (
+          properties.bucket && oldRecord.params[properties.bucket]
+        ) || adapter.bucket
+
+        if (oldKey && oldBucket && (oldKey !== key || oldBucket !== adapter.bucket)) {
+          await adapter.delete(oldKey, oldBucket)
+        }
+
+        return {
+          ...response,
+          record: record.toJSON(context.currentAdmin),
+        }
       }
     }
 
     return response
   }
 
-  const afterDelete = async (
+  const deleteFile = async (
     response: RecordActionResponse,
     request: ActionRequest,
     context: ActionContext,
   ): Promise<RecordActionResponse> => {
     const { record } = context
-    const s3Key = record?.param(properties.key)
-    if (s3Key) {
-      await removeFile(s3Key, credentials.bucket)
+
+    const key = record?.param(properties.key)
+
+    if (record && key) {
+      const storedBucket = properties.bucket && record.param(properties.bucket)
+      await adapter.delete(key, storedBucket || adapter.bucket)
     }
     return response
   }
 
+  const deleteFiles = async (
+    response: BulkActionResponse,
+    request: ActionRequest,
+    context: ActionContext,
+  ): Promise<BulkActionResponse> => {
+    const { records = [] } = context
+
+    await Promise.all(records.map(async (record) => {
+      const key = record?.param(properties.key)
+      if (record && key) {
+        const storedBucket = properties.bucket && record.param(properties.bucket)
+        await adapter.delete(key, storedBucket || adapter.bucket)
+      }
+    }))
+
+    return response
+  }
+
+  const fillRecordWithPath = async (record: RecordJSON): Promise<RecordJSON> => {
+    const key = record?.params[properties.key]
+    const storedBucket = properties.bucket && record?.params[properties.bucket]
+
+    if (key) {
+      // eslint-disable-next-line no-param-reassign
+      record.params[filePathProperty] = await adapter.path(key, storedBucket || adapter.bucket)
+    }
+
+    return record
+  }
+
+  const fillPath = async (response: RecordActionResponse): Promise<RecordActionResponse> => {
+    const { record } = response
+
+    return {
+      ...response,
+      record: await fillRecordWithPath(record),
+    }
+  }
+
+  const fillPaths = async (response: ListActionResponse): Promise<ListActionResponse> => {
+    const { records } = response
+
+    return {
+      ...response,
+      records: await Promise.all(records.map(fillRecordWithPath)),
+    }
+  }
+
+  const custom: PropertyCustom = {
+    fileProperty,
+    filePathProperty,
+    keyProperty: properties.key,
+    bucketProperty: properties.bucket,
+    mimeTypeProperty: properties.mimeType,
+    // bucket property can be empty so default bucket has to be passed
+    defaultBucket: adapter.bucket,
+    mimeTypes: validation?.mimeTypes,
+    maxSize: validation?.maxSize,
+  }
+
   const uploadFeature = buildFeature({
     properties: {
-      [properties.file]: {
```

Comment:
I removed that so it will be simpler - I use only one `virtual` property (fileProperty)